entry_test.go

```diff
@@ -43,17 +50,234 @@ func TestEntryPanicln(t *testing.T) {
+	// independent units, we covered all the obvious cases with one multi-part
+	// test. (Multi-part tests share an entry and test the entry state after
+	// each change.)
+	for _, independantTestCase := range [][]struct {
+		Group          Fields
+		ExpectedFields map[string][]string
+	}{
+		// This is the first (and currently only) independant test case:
+		{
+			// These are the multiple parts of the single test case:
+			{
+				Group: Fields{
+					"foo": "bar",
+				},
+				ExpectedFields: map[string][]string{
+					"foo": {"bar"},
+				},
+			},
+			{
+				Group: Fields{
+					"Fred":   func() {},
+					"George": func() {},
+					"Ron":    func() {},
+					"Ginnie": func() {},
+				},
+				ExpectedFields: map[string][]string{
+					"foo": {"bar"},
+					"logrus_error": {
+						`"can not add field \"Fred\""`,
+						`"can not add field \"George\""`,
+						`"can not add field \"Ron\""`,
+						`"can not add field \"Ginnie\""`,
+					},
+				},
+			},
+			{
+				Group: Fields{
+					"six": func() {},
+				},
+				ExpectedFields: map[string][]string{
+					"foo": {"bar"},
+					"logrus_error": {
+						`"can not add field \"Fred\""`,
+						`"can not add field \"George\""`,
+						`"can not add field \"Ron\""`,
+						`"can not add field \"Ginnie\""`,
+						`"can not add field \"six\""`,
+					},
+				},
+			},
+			{
+				Group: Fields{
+					"red": "green",
+				},
+				ExpectedFields: map[string][]string{
+					"foo": {"bar"},
+					"red": {"green"},
+					"logrus_error": {
+						`"can not add field \"Fred\""`,
+						`"can not add field \"George\""`,
+						`"can not add field \"Ron\""`,
+						`"can not add field \"Ginnie\""`,
+						`"can not add field \"six\""`,
+					},
+				},
+			},
+			{
+				Group: Fields{
+					"seven": func() {},
+				},
+				ExpectedFields: map[string][]string{
+					"foo": {"bar"},
+					"red": {"green"},
+					"logrus_error": {
+						`"can not add field \"Fred\""`,
+						`"can not add field \"George\""`,
+						`"can not add field \"Ron\""`,
+						`"can not add field \"Ginnie\""`,
+						`"can not add field \"six\""`,
+						`"can not add field \"seven\""`,
+					},
+				},
+			},
+		},
+	} {
+		// Independant tests have a new Entry, and a new expected final state.
+		entry := NewEntry(logger)
+		finalFieldsStr := ""
+		for _, fieldGroupTestPart := range independantTestCase {
+			// Set outBuffer as a new "file" to log to.
+			outBuffer := &bytes.Buffer{}
+			logger.Out = outBuffer
+
+			// ******** SYSTEM UNDER TEST ********
+			// Multiple calls to the WithFields() method is the primary
+			// thing being tested:
+			entry = entry.WithFields(fieldGroupTestPart.Group)
+			// Everything below here is analizing the Entry state after
+			// WithFields() is called:
+
+			// Capture logged output:
+			entry.Info("baz")
+			finalFieldsStr = outBuffer.String()
+
+			// Parse the output into key-value pairs.
+			keyValPairs := keyValPartsPattern.FindAllStringSubmatch(outBuffer.String(), -1)
```

Comment:
a json formatter will make the test less brittle here

Reply:
Looks very much the case... I'm a bit tangled up in another branch at the moment, but I **hope** to have something you can look at within the hour